tests/test_util.py

```diff
@@ -0,0 +1,180 @@
+
+
+def test_gentle_asarray_fits_rec_input():
+    cols = [fits.Column("col1", format="e", array=np.array([1, 2, 3, 4]))]
+    inp = fits.FITS_rec.from_columns(cols)
+    out_dtype = np.dtype([("col1", np.float32)])
+    result = util.gentle_asarray(inp, dtype=out_dtype)
+    assert result.dtype == out_dtype
+    assert_array_equal(result, inp)
+
+
+def test_gentle_asarray_fits_rec_pseudo_unsigned():
+    cols = [fits.Column("col1", format="j", array=np.array([1, 2, 3, 4], np.uint32), bzero=2147483648)]
+    inp = fits.FITS_rec.from_columns(cols)
+    out_dtype = np.dtype([("col1", np.uint32)])
+    result = util.gentle_asarray(inp, dtype=out_dtype)
+    assert result["col1"].dtype == np.uint32
+
+
+def test_gentle_asarray_nested_array():
+    in_dtype = np.dtype([("col1", np.dtype("2f")), ("col2", np.int16)])
+    inp = np.array([1, 2, 3, 4], dtype=in_dtype)
+    out_dtype = np.dtype([("col1", np.dtype("2f")), ("col2", np.int8)])
+    result = util.gentle_asarray(inp, dtype=out_dtype)
+    assert result.dtype == out_dtype
+    assert_array_equal(result["col2"], inp["col2"])
+
+
+def test_gentle_asarray_mismatched_column_names():
+    in_dtype = np.dtype([("col1", np.int16), ("col2", np.float32)])
+    inp = np.array([1, 2, 3, 4], dtype=in_dtype)
+    out_dtype = np.dtype([("col1", np.int16), ("foo", np.float32)])
+    with pytest.raises(ValueError):
+        util.gentle_asarray(inp, dtype=out_dtype)
+
+
+def test_gentle_asarray_column_name_case():
+    in_dtype = np.dtype([("col1", np.int16), ("col2", np.float32)])
+    inp = np.array([1, 2, 3, 4], dtype=in_dtype)
+    out_dtype = np.dtype([("COL1", np.int16), ("COL2", np.float32)])
+    result = util.gentle_asarray(inp, dtype=out_dtype)
+    assert result.dtype == out_dtype
+    assert_array_equal(result, inp)
+
+
+def test_gentle_asarray_scalar_input():
+    result = util.gentle_asarray(3.14159, dtype=np.float32)
+    assert result.dtype == np.float32
+    assert_array_equal(result, np.array([3.14159], dtype=np.float32))
+
+
+def test_gentle_asarray_invalid_conversion():
+    with pytest.raises(ValueError):
+        util.gentle_asarray(object(), dtype=np.float32)
+
+
+def test_get_short_doc():
+    assert util.get_short_doc({}) == ""
+    assert util.get_short_doc({"title": "Some schema title."}) == "Some schema title."
+    assert util.get_short_doc({
+        "title": "Some schema title.\nWhoops, another line."
+    }) == "Some schema title."
+    assert util.get_short_doc({
+        "title": "Some schema title.",
+        "description": "Some schema description.",
+    }) == "Some schema title."
+    assert util.get_short_doc({
+        "description": "Some schema description.",
+    }) == "Some schema description."
+    assert util.get_short_doc({
+        "description": "Some schema description.\nWhoops, another line.",
+    }) == "Some schema description."
+
+
+def test_ensure_ascii():
+    for inp in [b"ABCDEFG", "ABCDEFG"]:
+        util.ensure_ascii(inp) == "ABCDEFG"
```

Comment:
Is there a reason to have the `ensure_ascii` function anymore now that we are python3-only?  I.e., we probably shouldn't be passing around bytes as text, and if we are, it's probably because of python2.  But if it really requires `ascii`, then we should be making sure it actually is ascii, and it seems like this function only decodes bytes to ascii.  About unicode?

Reply:
That is odd, isn't it.  The code in fits_support.py is using this on values extracted from the schema, which will always be strings anyway.  The FITS standard restricts the characters that may be used in a header:

<img width="476" alt="image" src="https://user-images.githubusercontent.com/2895013/96302883-83dadf80-0fc7-11eb-8010-72f4c8d581d2.png">

So, maybe this method should actually be ensuring characters that fall within the allowed range?

Reply:
I'm going to punt on this by also moving it to fits_support.py, which I'm planning to add tests for in a future PR.